File: apps/api/src/digital-id-cards/digital-id-cards.controller.ts
import {
  Body,
  Controller,
  Get,
  Param,
  ParseUUIDPipe,
  Post,
  Query,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';
import { ApiBearerAuth, ApiOperation, ApiTags } from '@nestjs/swagger';
import { DigitalIdCardsService } from './digital-id-cards.service';
import { IssueCardDto } from './dto/issue-card.dto';
import { FreezeCardDto, ReissueCardDto, RevokeCardDto } from './dto/card-action.dto';
import { ListDigitalIdsQuery } from './dto/list-cards.dto';
import { DigitalIdAuthGuard } from '../auth/guards/digital-id-auth.guard';
import { RolesGuard } from '../auth/guards/roles.guard';
import { OfficerOnly } from '../auth/decorators/roles.decorator';
import { CurrentUser } from '../auth/decorators/current-user.decorator';
import { SijilliRequestUser } from '../auth/types';
import { Audit, AuditInterceptor } from '../audit/audit.interceptor';

@ApiTags('digital-id-cards')
@ApiBearerAuth()
@Controller('digital-id-cards')
@UseGuards(DigitalIdAuthGuard, RolesGuard)
@UseInterceptors(AuditInterceptor)
export class DigitalIdCardsController {
  constructor(private readonly cards: DigitalIdCardsService) {}

  @Get()
  @OfficerOnly('id_issuer', 'super_admin', 'auditor', 'registry_officer')
  @ApiOperation({ summary: 'List digital ID cards (cursor-paginated)' })
  list(@Query() q: ListDigitalIdsQuery, @CurrentUser() actor: SijilliRequestUser) {
    return this.cards.list(q, actor);
  }

  @Post('issue')
  @OfficerOnly('id_issuer', 'super_admin')
  @ApiOperation({ summary: 'Issue a new digital ID card (id_issuer only)' })
  @Audit({ action: 'issue_id', entity: 'digital_id_cards', entityIdFrom: 'card.id' })
  issue(@Body() dto: IssueCardDto, @CurrentUser() actor: SijilliRequestUser) {
    return this.cards.issue(dto, actor);
  }

  @Post(':id/freeze')
  @OfficerOnly('id_issuer', 'super_admin', 'registry_officer')
  @ApiOperation({ summary: 'Freeze a card (reversible)' })
  @Audit({ action: 'update', entity: 'digital_id_cards' })
  freeze(
    @Param('id', new ParseUUIDPipe()) id: string,
    @Body() dto: FreezeCardDto,
    @CurrentUser() actor: SijilliRequestUser,
  ) {
    return this.cards.freeze(id, dto, actor);
  }

  @Post(':id/revoke')
  @OfficerOnly('id_issuer', 'super_admin')
  @ApiOperation({ summary: 'Permanently revoke a card' })
  @Audit({ action: 'revoke_id', entity: 'digital_id_cards' })
  revoke(
    @Param('id', new ParseUUIDPipe()) id: string,
    @Body() dto: RevokeCardDto,
    @CurrentUser() actor: SijilliRequestUser,
  ) {
    return this.cards.revoke(id, dto, actor);
  }

  @Post(':id/reissue')
  @OfficerOnly('id_issuer', 'super_admin')
  @ApiOperation({ summary: 'Reissue a card (revokes old, mints fresh NFC keys)' })
  @Audit({ action: 'issue_id', entity: 'digital_id_cards', entityIdFrom: 'card.id' })
  reissue(
    @Param('id', new ParseUUIDPipe()) id: string,
    @Body() dto: ReissueCardDto,
    @CurrentUser() actor: SijilliRequestUser,
  ) {
    return this.cards.reissue(id, dto, actor);
  }
}

File: apps/api-dotnet/Controllers/DigitalIdCardsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Sarh.Api.Audit;
using Sarh.Api.Auth;
using Sarh.Api.Common;
using Sarh.Api.DigitalIdCards;

namespace Sarh.Api.Controllers;

[ApiController]
[Route("api/v1/digital-id-cards")]
[Authorize]
public class DigitalIdCardsController(DigitalIdCardsService cards) : ControllerBase
{
    [HttpGet]
    public Task<CursorPage<CardView>> List([FromQuery] ListCardsQuery q, CancellationToken ct)
        => cards.ListAsync(q, User.RequireUser(), ct);

    [HttpPost("issue")]
    [OfficerOnly("id_issuer", "super_admin")]
    [Audit(Action = AuditActions.IssueId, Entity = "digital_id_cards", EntityIdFrom = "card.id", CaptureResponseBody = false)]
    public Task<IssueCardResult> Issue([FromBody] IssueCardDto dto, CancellationToken ct)
        => cards.IssueAsync(dto, User.RequireUser(), ct);

    [HttpPost("{id:guid}/freeze")]
    [OfficerOnly("id_issuer", "super_admin", "registry_officer")]
    [Audit(Action = AuditActions.Update, Entity = "digital_id_cards")]
    public Task<CardView> Freeze(Guid id, [FromBody] FreezeCardDto dto, CancellationToken ct)
        => cards.FreezeAsync(id, dto, User.RequireUser(), ct);

    [HttpPost("{id:guid}/revoke")]
    [OfficerOnly("id_issuer", "super_admin")]
    [Audit(Action = AuditActions.RevokeId, Entity = "digital_id_cards")]
    public Task<CardView> Revoke(Guid id, [FromBody] RevokeCardDto dto, CancellationToken ct)
        => cards.RevokeAsync(id, dto, User.RequireUser(), ct);

    [HttpPost("{id:guid}/reissue")]
    [OfficerOnly("id_issuer", "super_admin")]
    [Audit(Action = AuditActions.IssueId, Entity = "digital_id_cards", EntityIdFrom = "card.id", CaptureResponseBody = false)]
    public Task<IssueCardResult> Reissue(Guid id, [FromBody] ReissueCardDto dto, CancellationToken ct)
        => cards.ReissueAsync(id, dto, User.RequireUser(), ct);

    [HttpPost("{id:guid}/reset-pin")]
    [OfficerOnly("id_issuer", "super_admin")]
    [Audit(Action = AuditActions.Update, Entity = "digital_id_cards", EntityIdFrom = "card_id", CaptureResponseBody = false)]
    public Task<ResetPinResult> ResetPin(Guid id, CancellationToken ct)
        => cards.ResetPinAsync(id, User.RequireUser(), ct);

    // Edit the card's validity window (expires_at). Other attributes are
    // immutable by design — see UpdateAsync.
    [HttpPatch("{id:guid}")]
    [OfficerOnly("id_issuer", "super_admin")]
    [Audit(Action = AuditActions.Update, Entity = "digital_id_cards")]
    public Task<CardView> Update(Guid id, [FromBody] UpdateCardDto dto, CancellationToken ct)
        => cards.UpdateAsync(id, dto, User.RequireUser(), ct);

    // Super-admin only. Soft-deletes the card (status=revoked, PIN+NFC
    // scrubbed) and purges nfc_card_secrets. Body is optional — if a
    // reason is supplied it lands in revoked_reason and id_issuance_history.
    [HttpDelete("{id:guid}")]
    [OfficerOnly("super_admin")]
    [Audit(Action = AuditActions.Delete, Entity = "digital_id_cards", EntityIdFrom = "card_id")]
    public Task<DeleteCardResult> Delete(Guid id, [FromBody] DeleteCardDto? dto, CancellationToken ct)
        => cards.DeleteAsync(id, dto ?? new DeleteCardDto(), User.RequireUser(), ct);
}
